src/interruptible/thread_data.h

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) 2015 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_INTERRUPTIBLE_THREAD_DATA_H
+#define BITCOIN_INTERRUPTIBLE_THREAD_DATA_H
+
+#include <atomic>
+#include <assert.h>
+#include <mutex>
+
+namespace interruptible
+{
+namespace detail
+{
+class thread_data
+{
+public:
+    thread_data() noexcept = default;
+    thread_data(thread_data&& t) = default;
+    thread_data& operator=(thread_data&& t) = default;
+    ~thread_data() = default;
+
+    thread_data(const thread_data&) = delete;
+    thread_data& operator=(const thread_data&) = delete;
+
+    template <typename T, typename... Args>
+    void start_thread(T&& func, Args&&... args)
+    {
+        m_thread = std::thread(std::forward<T>(func), std::forward<Args>(args)...);
+    }
+
+    void set_interrupted()
+    {
+        m_interrupted.store(true, std::memory_order_relaxed);
```

Comment:
Yeah, it's wild stuff. I don't really understand it either.

The synchronization effects of `notify` on non-`condition_variable` values aren't specified, so I think we have to be conservative. One approach would be a release fence after the store and an acquire load operation. There may be marginally better-performing options on the store side, but this has the benefit that acquire/release pairs are relatively comprehensible. The load could technically be relaxed and the compiler still "should" make sure thread that's doing the loads sees the update "in a reasonable amount of time", but the compiler can hoist relaxed loads out of loops if it sees fit and who knows what's a reasonable amount of time to a compiler. Acquire and release are both cheap on x86 and anyway, interrupt points are going to be places that we expect time consuming operations to be happening, not in tight loops.